### DISCUSSION TITLE
How to select fields to have more stable debugging scripts

### DISCUSSION THREAD

**Discussion by Pachat** [directus/directus#11817] [Archive]
With no id, we have to traverse the DOM. It is therefore very difficult to maintain debugging scripts.
For instance, if we change the order of fields, the DOM gets different and we have to adapt the debugging scripts, wasting a lot of time. 

It would be so much easier if we could rely on ids generated by Directus based on the field's name: 
it would keep debugging scripts working after many mere modifications.

This would also help to write specific CSS rules more easily in the Projects Settings

**Comment by Pachat** [comment]
After digging in the HTML generated by Directus, it appears that we have everything we need .
`<input loading="false" collection="mycollection" field="myfield_name" field-data="[object Object]" primary-key="+" autocomplete="off" type="number" step="1" data-v-497fc93a="" width="0">
`

Where Cypress generates code like
`cy.get(':nth-child(3) > .interface > .v-input > .input > input').type('1000000');`

we may just replace it manually by
`cy.get('input[collection="mycollection"][field="myfield_name"]').type('1000000');`


AS far as specific CSS rules in the Projects Settings are concerned, we may use
```
input[collection="mycollection"][field="myfield_name"] {
    ....
}
```

**Comment by rijkvanzanten** [comment]
Be aware: we don't currently consider the classes / ids used in the app breaking changes when making updates. These could change in any version at any time.

**Reply by Pachat** [reply]
Thank you for the precision.
However I suspect
`input[collection="mycollection"][field="myfield_name"]`
to stay for a long time.

**Reply by benhaynes** [reply]
Exactly. I think these selectors are **really** nice to provide, as long as it's clear in the docs that their updates aren't considered breaking changes.

**Reply by Pachat** [reply]
To complete the list of selectors, to type a `value` with Cypress in a WYSIWYG use:
```
    cy.get('div#myfield_name > div.tox.tox-tinymce > div.tox-editor-container > div.tox-sidebar-wrap > div.tox-edit-area ')
        .find('iframe').its('0.contentDocument.body').then(cy.wrap).type(value)

```
as you can see, here we use only the `myfield_name` not the `collection` for Directus does not provide the `collection` property.
This is OK as long as we have a unique `myfield_name`. But
- it would be nice to have the `collection` property in WYSIWYG: `collection="mycollection" ` 
```
<div id="resume" class="wysiwyg" autofocus="false" loading="false" width="full" type="text" collection="Projets"
     field-data="[object Object]" primary-key="3" data-v-6684db5c="" data-v-6f93e439=""><textarea....
```
--
Also, with DropDowns, we have `collection` and `fieldname` which is fine. We can use:
`div[collection="mycollection"][field="myfield_name"]`

But, with relations fields, like M2O, we don't get the `collection`, neither the `fieldname`:
```
    <div class="interface" data-v-6f93e439="" data-v-c7b47d1e=""><!---->
        <div class="many-to-one" autofocus="false" loading="false" width="half" type="integer"
             field-data="[object Object]" primary-key="3" data-v-1b0c186d="" data-v-6f93e439="">...
```
- It would be nice to have the `collection` and `field name` properties for relations fields. 
That would give more to have more stable debugging scripts0
